### PR TITLE
Proposed version 2.0 - file streaming, version tagging

### LINKED ISSUE #34: Version Tagging
This is a nice-to-have in case anyone wants to encrypt with version 1.x of this library in 2015 then decrypt it with version 3.x in 2025.

### LINKED ISSUE #39: Support API for Encrypting/Decrypting files.
You can't encrypt super-large files with this library because you have to load them into RAM. We could either (1) Have a "stream" sort of API, where you feed it data and it feeds back ciphertext, but there are problems with this, e.g. when decrypting you'll get all of the plaintext and then only the last call will check the MAC, but the damage has already been done, or (2) Support file encryption/decryption only, so that we can encapsulate all of that.

### PR DESCRIPTION
**Pull request by paragonie-scott**
Satisfies the proposal outlined in #71 

Supersedes #63 
Closes #34 
Closes #39

### CODE REVIEW

File: src/Crypto.php [hunk 1]
```diff
@@ -62,10 +45,10 @@
      */
     public static function createNewRandomKey()
     {
-        self::runtimeTest();
-        return self::secureRandom(self::KEY_BYTE_SIZE);
+        $config = self::getVersionConfig(self::VERSION);
+        return self::secureRandom($config['KEY_BYTE_SIZE']);
     }
-
+
```
Comment (sstok) [inline src/Crypto.php]: Trailing whitespace ;)

**Comment by sstok** [comment]
https://github.com/paragonie/php-encryption/blob/files/src/Crypto.php#L322
`if (\in_array(self::CIPHER_METHOD, \openssl_get_cipher_methods()) === FALSE) {` was this left unchanged intentionally?

File: src/Crypto.php [hunk 2]
```diff
@@ -253,22 +353,30 @@ public static function runtimeTest()
...
             $plaintext,
-            self::CIPHER_METHOD,
+            $config['CIPHER_METHOD'],
             $key,
             OPENSSL_RAW_DATA,
             $iv
         );
 
         if ($ciphertext === false) {
+            \var_dump([
```
Comment (sstok) [inline src/Crypto.php]: Debug code

File: src/Crypto.php [hunk 3]
```diff
@@ -409,39 +432,21 @@ private static function HKDF($hash, $ikm, $length, $info = '', $salt = null)
...
-        $blind = self::createNewRandomKey();
-        $message_compare = \hash_hmac(self::HASH_FUNCTION, $message_hmac, $blind);
-        $correct_compare = \hash_hmac(self::HASH_FUNCTION, $correct_hmac, $blind);
-        return $correct_compare === $message_compare;
+        $message_hmac = \hash_hmac($config['HASH_FUNCTION'], $message, $key, true);
+        return self::hashEquals($correct_hmac, $message_hmac);
     }
-
-    private static function testEncryptDecrypt()
+
```
Comment (sstok) [inline src/Crypto.php]: Trailing whitespace ;)

File: src/Crypto.php [hunk 4]
```diff
@@ -516,13 +524,18 @@ private static function HKDFTestVector()
...
-        $okm = self::hexToBytes(
+        $okm = self::hexToBin(
             "2c91117204d745f3500d636a62f64f0a" .
             "b3bae548aa53d423b0d1f27ebba6f5e5" .
             "673a081d70cce7acfc48"
         );
-        $computed_okm = self::HKDF("sha1", $ikm, $length);
+        $computed_okm = self::HKDF("sha1", $ikm, $length, '', null, $config);
         if ($computed_okm !== $okm) {
+            \var_dump([
```
Comment (sstok) [inline src/Crypto.php]: Debug code

File: src/File.php [hunk 5]
```diff
@@ -0,0 +1,762 @@
...
+             */
+            if(\fseek($inputHandle, (-1 * $config['MAC_BYTE_SIZE']), SEEK_END) === false) {
+                throw new Ex\CannotPerformOperation(
+                    'Cannot seek to beginning of MAC within input file'
+                );
+            }
+            
+            // Grab our last position of ciphertext before we read the MAC
+            $cipher_end = \ftell($inputHandle) - 1;
+            if ($cipher_end === false) {
```
Comment (sstok) [inline src/File.php]: This is going to fail as with the `-1` its converted to an integer which doesn't strict equal false!

File: tests/stream/decrypt.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
+<?php
+require_once \dirname(\dirname(__DIR__)).'/autoload.php';
+
+if (!\file_exists('key.txt')) {
+    echo 'You need to generate an key first!', "\n";
```
Comment (sstok) [inline tests/stream/decrypt.php]: *a* key (not *an*) :)

**Comment by paragonie-scott** [comment]
Ugh, wow, I can't believe I forgot to remove my debug code. Stupid, stupid, stupid.

Commit incoming.

**Comment by paragonie-scott** [comment]
The latest test verifies that our nonce incrementing strategy is functioning correctly. Aside from the danger of birthday collisions, nonce collisions are mitigated.

Mitigating nonce reuse is nontrivial, since we're limited to 128 bits. I think we would first need to create a subkey for each file and store the random HKDF or PBKDF2 salt in the file data to make this impractical.

**Comment by paragonie-scott** [comment]
Since HKDF accepts a salt parameter, I added a random per-file salt to the file encrypted file header in addition to the IV.

Without the salt, the odds of a birthday collision after 2^64 files is approximately 50%. Depending on the file size, a nonce reuse becomes a real problem.

Using a random salt and a nonrandom nonce (e.g. all 0's) is virtually identical to the existing situation.

Using a random salt AND a random nonce raises our birthday resistance to require 2^128 files for a 50% chance of birthday collision. (This is effectively similar to a 256-bit nonce.)

**Comment by paragonie-scott** [comment]
The lastest commit should satisfy #93

**Comment by paragonie-scott** [comment]
Forgot about the runtime tests. Need new test vectors.

File: doc/02-Crypto.php.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
+Symmetric Key Encryption
+========================
+
+At a glance:
+
+* **Cipher and Mode**: `AES-128-CBC`
```
Comment (defuse) [inline doc/02-Crypto.php.md]: This is now CTR.

File: doc/02-Crypto.php.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+* **Padding**: `PKCS#7`
```
Comment (defuse) [inline doc/02-Crypto.php.md]: Not used since CTR.

**Comment by paragonie-scott** [comment]
One thing that the `File` class doesn't have, but the `Crypto` class does, is the runtime tests. Would you prefer that they be run in that instance as well? (I'd wager your answer is, "Yes".)

File: doc/02-Crypto.php.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+```php
+try {
+    $key = \Defuse\Crypto\Crypto::createNewRandomKey();
+    // WARNING: Do NOT encode $key with bin2hex() or base64_encode(),
+    // they may leak the key to the attacker through side channels.
+    // 
+    // This library offers these two for cache-timing-safe encoding:
+    //
+    //     \Defuse\Crypto\Crypto\binToHex()
+    //     \Defuse\Crypto\Crypto\hexToBin()
```
Comment (defuse) [inline doc/02-Crypto.php.md]: Awesome!

File: doc/02-Crypto.php.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    // Either:
+    //   1. The ciphertext was modified by the attacker,
+    //   2. The key is wrong, or
+    //   3. $ciphertext is not a valid ciphertext or was corrupted.
+    // Assume the worst.
+    die('DANGER! DANGER! The ciphertext has been tampered with!');
+} catch (CryptoTestFailedException $ex) {
+    die('Cannot safely perform decryption');
+} catch (CannotPerformOperationException $ex) {
+    die('Cannot safely perform decryption');
```
Comment (defuse) [inline doc/02-Crypto.php.md]: An important property to keep in mind: Tampering the ciphertext should *only* be able to lead to InvalidCiphertext exceptions, i.e. the attacker shouldn't be able to trick you into thinking the runtime test failed or something like that (just thinking out loud).

File: doc/02-Crypto.php.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+}
+```
+
+Decrypting a Message
+--------------------
+
+```php
+try {
+    $decrypted = self::Decrypt($ciphertext, $key);
+} catch (\Defuse\Crypto\Exception\InvalidCiphertext $ex) { // VERY IMPORTANT
```
Comment (defuse) [inline doc/02-Crypto.php.md]: This exception seems to be in a different namespace than the ones below, is it still correct?

File: doc/02-Crypto.php.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+try {
+    $decrypted = self::legacyDecrypt($ciphertext, $key);
+} catch (\Defuse\Crypto\Exception\InvalidCiphertext $ex) { // VERY IMPORTANT
+    // Either:
+    //   1. The ciphertext was modified by the attacker,
+    //   2. The key is wrong, or
+    //   3. $ciphertext is not a valid ciphertext or was corrupted.
+    // Assume the worst.
+    die('DANGER! DANGER! The ciphertext has been tampered with!');
+} catch (CryptoTestFailedException $ex) {
```
Comment (defuse) [inline doc/02-Crypto.php.md]: I think these might be in a different namespace now?

File: src/Core.php [hunk 13]
```diff
@@ -0,0 +1,394 @@
+<?php
+namespace Defuse\Crypto;
+
+use \Defuse\Crypto\Exception as Ex;
+
+class Core
+{
+    const VERSION = "\xD3\xF5\x02\x00";
```
Comment (defuse) [inline src/Core.php]: What is the meaning of these bytes? What will the next version number be?

File: src/File.php [hunk 14]
```diff
@@ -0,0 +1,742 @@
...
+final class File extends Core implements StreamInterface
+{
+    /**
+     * Use this to generate a random encryption key.
+     *
+     * @return string
+     */
+    public static function createNewRandomKey()
+    {
+        $config = self::getVersionConfig(self::VERSION);
```
Comment (defuse) [inline src/File.php]: Do file ciphertexts start with the same 'version' as normal ciphertexts? There should be a way to tell them apart using the version number (or something) and an exception thrown if you try to decrypt with the wrong API (e.g. normal ciphertext saved to a file).

File: src/File.php [hunk 15]
```diff
@@ -0,0 +1,742 @@
...
+                    'Cannot open input file for encrypting'
+                );
+            }
+            \stream_set_read_buffer($if, 0);
+
+            /**
+             * Output file handle
+             */
+            $of = \fopen($outputFilename, 'wb');
+            if ($of === false) {
```
Comment (defuse) [inline src/File.php]: `$if` doesn't get closed here.

File: src/File.php [hunk 16]
```diff
@@ -0,0 +1,742 @@
...
+                    'Cannot open input file for decrypting'
+                );
+            }
+            \stream_set_read_buffer($if, 0);
+
+            /**
+             * Output file handle
+             */
+            $of = \fopen($outputFilename, 'wb');
+            if ($of === false) {
```
Comment (defuse) [inline src/File.php]: `$if` doesn't get closed here.

File: src/File.php [hunk 17]
```diff
@@ -0,0 +1,742 @@
...
+     *
+     * @param string $header
+     */
+    protected static function getVersionConfig($header)
+    {
+        $valid = 0;
+        $valid |= $header[0] ^ "\xDE";
+        $valid |= $header[1] ^ "\xF5";
+        $major = \ord($header[2]);
+        $minor = \ord($header[3]);
```
Comment (defuse) [inline src/File.php]: Ah, here's the code that defines what the version means.

**Comment by defuse** [comment]
I haven't looked too closely at the issues, but I'm going to merge this into a v2.0 branch so that it's easier to work with, and then open tickets for the issues I found.

File: src/File.php (hunk 14 above)
Comment (paragonie-scott) [inline src/File.php]: That's a good idea. Would you be fine with the last bit of the second byte being used to specify file vs string?

File: src/Core.php (hunk 13 above)
Comment (paragonie-scott) [inline src/Core.php]: Either `"\xD3\xF5\x02\x01"` or `"\xD3\xF5\x03\x00"` depending on how the library changes over time. :)

File: doc/02-Crypto.php.md (hunk 12 above)
Comment (paragonie-scott) [inline doc/02-Crypto.php.md]: Yes! Good catch.

**Comment by paragonie-scott** [comment]
Okay, sounds good to me. Need me to resubmit the PR?

**Comment by defuse** [comment]
@paragonie-scott: I just created the v2.0 branch, so feel free to submit PRs to it if you feel like fixing any of those things. :)

**Comment by defuse** [comment]
Hang on, I'm still trying to figure out how to push that branch to github.

**Comment by defuse** [comment]
There we go.

**Comment by paragonie-scott** [comment]
This has been merged in another branch.

File: src/File.php (hunk 14 above)
Comment (defuse) [inline src/File.php]: I opened #105 for that. Yes. So it would be DEF4 and DEF5?

File: src/File.php (hunk 17 above)
Comment (defuse) [inline src/File.php]: Hey, it's "DE" here but "D3" up there! How is it working!?

**Pull request closed without merging**